File: content/code/post52/fp_decode.hs
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-12.7
  --package containers
-}


import Control.Monad (forM_)
import Data.IntMap.Strict (IntMap, fromList, (!?))
import System.Environment (getArgs)


data InterpreterResult = NotInterpreted
                       | InterpreterOk String
                       | InterpreterError String

type MessageID = Int
type MessageStatus = Int

newtype Interpreter = Interpreter { getInterpreter :: MessageID -> MessageStatus -> InterpreterResult }


createMotorInterpreter :: Int -> String -> Interpreter
createMotorInterpreter motorID motorName =
  let
    motorNetworkID = motorID + 20
    statusMap = fromList [ (0, "Not moving")
                         , (1, "Moving")
                         , (2, "Hardware defect") ]
    interpreter mID mStatus =
      if (mID == motorNetworkID)
      then
        case (statusMap !? mStatus) of
          Nothing -> InterpreterError ("Unknown status " ++ (show mStatus) ++ " for motor " ++ motorName)
          Just s -> InterpreterOk ("Motor " ++ motorName ++ ": " ++ s)
      else
        NotInterpreted
  in
    Interpreter interpreter


createSensorInterpreter :: Int -> String -> Interpreter
createSensorInterpreter sensorID sensorName =
  let
    sensorNetworkID = sensorID + 40
    statusMap = fromList [ (0, "Ok")
                         , (1, "Above threshold")
                         , (2, "Below threshold") ]
    interpreter sID sStatus =
      if (sID == sensorNetworkID)
      then
        case (statusMap !? sStatus) of
          Nothing -> InterpreterError ("Unknown status " ++ (show sStatus) ++ " for sensor " ++ sensorName)
          Just s -> InterpreterOk ("Sensor " ++ sensorName ++ ": " ++ s)
      else
        NotInterpreted
  in
    Interpreter interpreter


createGpioInterpreter :: Int -> String -> Interpreter
createGpioInterpreter gpioID gpioName =
  let
    gpioNetworkID = gpioID + 50
    statusMap = fromList [ (0, "Off")
                         , (1, "On") ]
    interpreter gID gStatus =
      if (gID == gpioNetworkID)
      then
        case (statusMap !? gStatus) of
          Nothing -> InterpreterError ("Unknown status " ++ (show gStatus) ++ " for gpio " ++ gpioName)
          Just s -> InterpreterOk ("GPIO " ++ gpioName ++ ": " ++ s)
      else
        NotInterpreted
  in
    Interpreter interpreter


deviceAInterpreterFunc deviceID deviceStatus =
  let
    statusMap = fromList [ (0, "Idle")
                         , (1, "Processing")
                         , (2, "Sending") ]
  in
    if (deviceID == 2)
    then
      case (statusMap !? deviceStatus) of
        Nothing -> InterpreterError ("Unknown status " ++ (show deviceStatus) ++ " for Device_A")
        Just s -> InterpreterOk ("Device_A: " ++ s)
    else
      NotInterpreted

deviceAInterpreter = Interpreter deviceAInterpreterFunc


deviceBInterpreterFunc deviceID deviceStatus =
  let
    statusMap = fromList [ (0, "Idle")
                         , (1, "Calculating") ]
  in
    if (deviceID == 4)
    then
      case (statusMap !? deviceStatus) of
        Nothing -> InterpreterError ("Unknown status " ++ (show deviceStatus) ++ " for Device_B")
        Just s -> InterpreterOk ("Device_B: " ++ s)
    else
      NotInterpreted

deviceBInterpreter = Interpreter deviceBInterpreterFunc


motors = [ (1, "Motor_A")
         , (2, "Motor_B")
         , (3, "Motor_C")]


sensors = [ (1, "Sensor_A")
          , (2, "Sensor_B")]


gpios = [ (1, "GPIO_A")
        , (2, "GPIO_B")]


interpreters = [createMotorInterpreter  i n | (i,n) <- motors ] ++
               [createSensorInterpreter i n | (i,n) <- sensors] ++
               [createGpioInterpreter   i n | (i,n) <- gpios  ] ++
               [deviceAInterpreter, deviceBInterpreter]


decodeWithInterpreters :: [Interpreter] -> MessageID -> MessageStatus -> InterpreterResult
decodeWithInterpreters interpreters id status =
  let tryInterpreter (Interpreter curr) next =
        case (curr id status) of
          NotInterpreted -> next
          result -> result
  in foldr tryInterpreter NotInterpreted interpreters


decode = decodeWithInterpreters interpreters


onOk :: String -> IO ()
onOk s = putStrLn s

onErr :: String -> IO ()
onErr s = putStrLn ("Error! " ++ s)


handleResult err ok res =
  case res of
    NotInterpreted -> return ()
    InterpreterOk s -> ok s
    InterpreterError s -> err s


main = do
  fileName <- head <$> getArgs
  fileContent <- readFile fileName
  let messages = map read (lines fileContent) :: [(Int,Int)]
      printResult = handleResult onErr onOk
  forM_ messages (printResult . (uncurry decode))

File: content/code/post52/fp_decode.py
def create_group_interpreter(range_start, range_end, status_map, group_name):
    def interpreter(ID, status):
        if range_start <= ID < range_end:
            device_nr = ID - range_start
            status_string = status_map.get(status, None)
            if status_string is None:
                return None
            else:
                return (f"{group_name} #{device_nr}: {status_string}")
        return None
    return interpreter


motor_status_map = {
    0: "Not moving",
    1: "Moving",
    2: "Hardware defect",
}

sensor_status_map = {
    0: "Ok",
    1: "Above threshold",
    2: "Below threshold",
}

gpio_status_map = {
    0: "Off",
    1: "On",
}


motor_interpreter = create_group_interpreter(20, 30, motor_status_map, "Motor")
sensor_interpreter = create_group_interpreter(40, 50, sensor_status_map, "Sensor")
gpio_interpreter = create_group_interpreter(50, 58, gpio_status_map, "GPIO")


def device_a_interpreter(ID, status):
    if ID == 2:
        status_map = {
            0: "Idle",
            1: "Processing",
            2: "Sending",
        }
        status_string = status_map.get(status, None)
        if status_string is None:
            return None
        return (f"Device_A: {status_string}")
    return None


def device_b_interpreter(ID, status):
    if ID == 4:
        status_map = {
            0: "Idle",
            1: "Calculating",
        }
        status_string = status_map.get(status, None)
        if status_string is None:
            return None
        return (f"Device_B: {status_string}")
    return None

# Main application code starts here

interpreters = [motor_interpreter,
                sensor_interpreter,
                gpio_interpreter,
                device_a_interpreter,
                device_b_interpreter]


def decode(ID, status):
    for interpreter in interpreters:
        interpretation = interpreter(ID, status)
        if interpretation is not None:
            return interpretation
    return None


if __name__ == '__main__':
    import sys
    import re

    message_re = re.compile(r'\((?P<ID>\d+),(?P<status>\d+)\)')
    protocol_dump_file = sys.argv[1]

    def unknown(ID, status):
        print(f"Unknown ID {ID} with status {status}")

    with open(protocol_dump_file, 'r') as protocol_dump:
        for line in protocol_dump.readlines():
            m = message_re.match(line)
            if m:
                ID = int(m.group('ID'))
                status = int(m.group('status'))
                decode_result = decode(ID, status)
                if decode_result is not None:
                    print(decode_result)
                else:
                    unknown(ID, status)
